### PR TITLE
Add protobuf 3.5.1.1 version

### PR DESCRIPTION
**Pull request by guillaume-michel**
Specify library name and version:  **protobuf/3.5.1.1**

This library is a dependency for Kinova robot's proprietary library.

---

- [x] I've read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md).
- [x] I've used a [recent](https://github.com/conan-io/conan/releases/latest) Conan client version close to the [currently deployed](https://github.com/conan-io/conan-center-index/blob/master/.c3i/config_v1.yml#L6).
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by prince-chrismc** [comment]
You need to wait for access request before it will be merged :)

**Review by prince-chrismc — APPROVED** [review]

**Review by jcar87 — CHANGES_REQUESTED** [review]
Hi @guillaume-michel, thanks for opening this PR.

We typically don't add older versions unless [sufficiently justified](https://github.com/conan-io/conan-center-index/blob/master/docs/adding_packages/sources_and_patches.md#supported-versions), and Protobuf 3.5.1.1 is a little over 3 years old.

We have recently removed logic in the recipe to support older versions, which simplifies the logic in the recipe. This would be moving in the opposite directions, and may introduce corner cases in which certain configurations don't work. 

In particular with protobuf, my understanding is that Protobuf 3 is backwards compatible and for users it should generally be safe to upgrade to newer versions without experiencing breakage - so long as they have the ability to rebuild/relink protobuf-dependent code (as there are no guarantees on binary compatibility).

Is that the case here? Would love to understand what's holding back being able to move to a newer version!

**Comment by prince-chrismc** [comment]
We can put the older version in a separate recipe folder so not complicate the newer ones

**Comment by guillaume-michel** [comment]
> Hi @guillaume-michel, thanks for opening this PR.
> 
> We typically don't add older versions unless [sufficiently justified](https://github.com/conan-io/conan-center-index/blob/master/docs/adding_packages/sources_and_patches.md#supported-versions), and Protobuf 3.5.1.1 is a little over 5 years old.
> 
> We have recently removed logic in the recipe to support older versions, which simplifies the logic in the recipe. This would be moving in the opposite direction, and may introduce corner cases in which certain configurations don't work.
> 
> In particular with protobuf, my understanding is that Protobuf 3 is backwards compatible and for users it should generally be safe to upgrade to newer versions without experiencing breakage - so long as they have the ability to rebuild/relink protobuf-dependent code (as there are no guarantees on binary compatibility).
> 
> Is that the case here? Would love to understand what's holding back being able to move to a newer version, so that we can take appropriate action, as we'd love to support this use case :)

Hi,
Thanks for taking some time to review the PR.

We are building a medical robot which is operated via a vendor provided **binary only library which exposes protobuf 3.5.1.1 in its public interface**. We tried to ask for an upgraded version but did not work out in the end. Our only option right now is to use protobuf 3.5.1.1.

In general, I agree with removing old legacy code but in practice it is not always possible. I am also worried that 2 years from now I could not find the dependencies I am looking for in conan and could not build old version. I am aware that I can use my own artifactory to keep my binaries but how am I supposed to maintain recipes which disappears from conan as time goes on? My naive approach would be to keep a compulsive cumulative strategy but I know it is a large burden to maintain.

**Comment by prince-chrismc** [comment]
> but how am I supposed to maintain recipes which disappears from conan as time goes on?

You can see our FAQ for more details but we never delete anything from ConanCenter. That does mean version stop getting updates. Unless there's a specific use-case (which this does seem to meet) then we drop support them when they are too burdensome.

The packages are always available but the support and maintenance is not 😸 hopeful that clarifies the policy

***

If you contribute a working version that is not too intrusive I dont see why it could not be accepted but thank you for filling in the explanation as to why ❤️

**Comment by jcar87** [comment]
> > Hi @guillaume-michel, thanks for opening this PR.
> > We typically don't add older versions unless [sufficiently justified](https://github.com/conan-io/conan-center-index/blob/master/docs/adding_packages/sources_and_patches.md#supported-versions), and Protobuf 3.5.1.1 is a little over 5 years old.
> > We have recently removed logic in the recipe to support older versions, which simplifies the logic in the recipe. This would be moving in the opposite direction, and may introduce corner cases in which certain configurations don't work.
> > In particular with protobuf, my understanding is that Protobuf 3 is backwards compatible and for users it should generally be safe to upgrade to newer versions without experiencing breakage - so long as they have the ability to rebuild/relink protobuf-dependent code (as there are no guarantees on binary compatibility).
> > Is that the case here? Would love to understand what's holding back being able to move to a newer version, so that we can take appropriate action, as we'd love to support this use case :)
> 
> Hi, Thanks for taking some time to review the PR.
> 
> We are building a medical robot which is operated via a vendor provided **binary only library which exposes protobuf 3.5.1.1 in its public interface**. We tried to ask for an upgraded version but did not work out in the end. Our only option right now is to use protobuf 3.5.1.1.
> 
> In general, I agree with removing old legacy code but in practice it is not always possible. I am also worried that 2 years from now I could not find the dependencies I am looking for in conan and could not build old version. I am aware that I can use my own artifactory to keep my binaries but how am I supposed to maintain recipes which disappears from conan as time goes on? My naive approach would be to keep a compulsive cumulative strategy but I know it is a large burden to maintain.

Hi @guillaume-michel - thanks for your reply. Indeed I can see your predicament. If you are a consuming closed-source binaries that impose a specific version, this could be a problem. Out of curiosity, are the Protobuf binaries also provided with the Kinova library? Given that Protobuf does not guarantee binary compatibility (unlike Qt for example, that has some guarantees), even if you independently build it I'm not sure it's guaranteed to work. 

As @prince-chrismc  points out, it should be clarified that the Conan Center remote will serve all versions that have ever been available - even if the recipe in the repository has stopped supporting older versions, so if Protobuf 3.5 were to be published, it would be available in the future even if the recipe stops supporting it.

As he mentions, if we can get this to pass our CI checks on the expected platforms, we're happy to proceed with this PR.

**Comment by guillaume-michel** [comment]
I am trying to fix the CI failures for protobuf/3.5.1.1 with shared library but I don't know how to fix it.

profile failure1 is:
```
[settings]
arch=x86_64
build_type=Debug
compiler=gcc
compiler.libcxx=libstdc++
compiler.version=5
os=Linux
[options]
protobuf:shared=True
```

When I am building the package locally on my machine with the following command:
`conan create . 3.5.1.1@ --build -pr=failure1`  which is the command executed in the CI,
I get package ID: 75ad3442f6ede2a299883b25274d1968d8cbd66d 
And I have the same failure as in the CI: protoc doesn't find libprotocd.so but it exists in the lib folder.

When I am building the package with the following command:
`conan create . 3.5.1.1@ --build -pr:h=failure1 -pr:b=failure1` 
I get the same package ID: 75ad3442f6ede2a299883b25274d1968d8cbd66d
But this time no errors.

Repeatting the same with 3.21.9 work in both cases.

Comparing protoc for both versions shows that protoc 3.5.1.1 is missing:
```
 (RUNPATH)            Library runpath: [$ORIGIN/../lib]
``` 
which is there because there is the following in the RUNPATH in libprotocd.so v3.21.9:
```
(RUNPATH)            Library runpath: [$ORIGIN]
```

Any idea why the test_package fails when using a single profile?
Any idea why it is working with dual profile without the correct RUNPATH?

Thanks for your help

**Comment by guillaume-michel** [comment]
In the meantime, I will have a look at the 3.5.1.1 CMakeLists.txt and try to patch the file so the RUNPATH is correctly added.

**Comment by guillaume-michel** [comment]
> > > Hi @guillaume-michel, thanks for opening this PR.
> > > We typically don't add older versions unless [sufficiently justified](https://github.com/conan-io/conan-center-index/blob/master/docs/adding_packages/sources_and_patches.md#supported-versions), and Protobuf 3.5.1.1 is a little over 5 years old.
> > > We have recently removed logic in the recipe to support older versions, which simplifies the logic in the recipe. This would be moving in the opposite direction, and may introduce corner cases in which certain configurations don't work.
> > > In particular with protobuf, my understanding is that Protobuf 3 is backwards compatible and for users it should generally be safe to upgrade to newer versions without experiencing breakage - so long as they have the ability to rebuild/relink protobuf-dependent code (as there are no guarantees on binary compatibility).
> > > Is that the case here? Would love to understand what's holding back being able to move to a newer version, so that we can take appropriate action, as we'd love to support this use case :)
> > 
> > 
> > Hi, Thanks for taking some time to review the PR.
> > We are building a medical robot which is operated via a vendor provided **binary only library which exposes protobuf 3.5.1.1 in its public interface**. We tried to ask for an upgraded version but did not work out in the end. Our only option right now is to use protobuf 3.5.1.1.
> > In general, I agree with removing old legacy code but in practice it is not always possible. I am also worried that 2 years from now I could not find the dependencies I am looking for in conan and could not build old version. I am aware that I can use my own artifactory to keep my binaries but how am I supposed to maintain recipes which disappears from conan as time goes on? My naive approach would be to keep a compulsive cumulative strategy but I know it is a large burden to maintain.
> 
> Hi @guillaume-michel - thanks for your reply. Indeed I can see your predicament. If you are a consuming closed-source binaries that impose a specific version, this could be a problem. Out of curiosity, are the Protobuf binaries also provided with the Kinova library? Given that Protobuf does not guarantee binary compatibility (unlike Qt for example, that has some guarantees), even if you independently build it I'm not sure it's guaranteed to work.
> 
> As @prince-chrismc points out, it should be clarified that the Conan Center remote will serve all versions that have ever been available - even if the recipe in the repository has stopped supporting older versions, so if Protobuf 3.5 were to be published, it would be available in the future even if the recipe stops supporting it.
> 
> As he mentions, if we can get this to pass our CI checks on the expected platforms, we're happy to proceed with this PR.

@jcar87 Kinova provides 2 libraries for their robot:
- 1 with including protobuf 3.5.1.1
- 1 not including protobuf but requires to link with an external library.

We chose to use the 1 without included protobuf because we are using protobuf elsewhere in our codebase and we don't want to link with whole kinova library for that part. Hope it clarifies the situation.

File: recipes/protobuf/all/patches/upstream-pr-4709-remove-js_embed.patch
Comment (SpaceIm) [inline recipes/protobuf/all/patches/upstream-pr-4709-remove-js_embed.patch:40]: Why don't you keep this file if it is not used anymore? Patch would be simplier.

File: recipes/protobuf/all/patches/upstream-pr-4620-rpath-management.patch
Comment (SpaceIm) [inline recipes/protobuf/all/patches/upstream-pr-4620-rpath-management.patch:49]: It can avoided, test_v1_package should be fixed instead:

```python
from conans import ConanFile, CMake, tools
import os


class TestPackageConan(ConanFile):
    settings = "os", "arch", "compiler", "build_type"
    generators = "cmake", "cmake_find_package_multi"
    test_type = "explicit"

    def requirements(self):
        self.requires(self.tested_reference_str)

    def build_requirements(self):
        if hasattr(self, "settings_build") and cross_building(self):
            self.build_requires(self.tested_reference_str)

    def build(self):
        with tools.no_op() if hasattr(self, "settings_build") and cross_building(self) else tools.run_environment(self):
            cmake = CMake(self)
            cmake.definitions["protobuf_LITE"] = self.options["protobuf"].lite
            cmake.configure()
            cmake.build()

    def test(self):
        if not tools.cross_building(self):
            self.run("protoc --version", run_environment=True)
            self.run(os.path.join("bin", "test_package"), run_environment=True)
```

**Pull request closed without merging**
